### ISSUE TITLE
Spring Arm Node

### ISSUE THREAD

**Issue by CiaraBurkett** [godotengine/godot#16508]
**Godot version:** 3.0

**Issue description:** Spring Arm Node

For example, Unreal Engine has a [Spring Arm Component](https://docs.unrealengine.com/latest/INT/Gameplay/HowTo/UsingCameras/SpringArmComponents/) you can use to attach to cameras. It's extremely useful, and it seems like it'd be a great match for making Godot that much easier/more fun to use.

![Image of Yaktocat](https://docs.unrealengine.com/latest/images/Gameplay/HowTo/UsingCameras/SpringArmComponents/Spring4.jpg)

**Comment by vnen** [comment]
I prefer you just describe what you want instead of linking to the documentation of another engine. Different engines have different workflows and it's not always straightforward to understand how their feature would apply to Godot. Also, the image is not helpful, it's just a bunch of settings which are meaningless without a description.

**Comment by LikeLakers2** [comment]
What vnen said. This issue left me a bit confused, and admittedly I've not been able to understand UE4's documentation too well. It'd be best if you try to explain what this idea does, even if you're pretty much rewording the documentation you linked to.

Reading up on the docs, from what I can understand, this causes the camera to have a slight (configurable) lag behind the player character (hence springy). It would also prevent the camera from going through collision. If I'm to assume this is what you're suggesting, and this is not already implemented, then I would be in support of such a feature.

**Comment by Zireael07** [comment]
The major benefit of spring arm is the camera not going through collisions.

**Comment by Akira1San** [comment]
Having an easy build-in setups is a plus.

**Comment by QbieShay** [comment]
How would you expect it to work?
Would you want it to be a built-in Node or a plugin with a scene and a script?

**Comment by mhilbrunner** [comment]
UE4's spring arm component is "**springy**" in the sense that it tries to keep the distance (of the attached thing/camera) as close to the configured maximum distance as possible, but **compresses** and **expands** according to the environment:

When you move the camera such that a wall would be between the thing it follows (generally the player) it compresses so the camera is between the player and that wall.

And yeah, it smoothes/lags movement a bit.

Short, few second-long demo:
https://www.youtube.com/watch?v=-hcoNgtoT8I

**Comment by CiaraBurkett** [comment]
@vnen Edited for clarity. Also, the attached image was to show a few of the available settings, not to explain the component (that's what the link was originally for).

**Comment by QbieShay** [comment]
@CiaraBurkett @mhilbrunner would you prefer a dedicated node or a property of the camera?
My idea at the moment is to create a special physic node that behaves like that, with some options about rotation inheritance, elasticity, collisions, etc. Then the camera could be put as a child of this node. WDYT?
I could work on this.

**Comment by Noshyaar** [comment]
@QbieShay a dedicated node for it would be more flexible than bundling it with a camera I think.

**Comment by QbieShay** [comment]
@Noshyaar yes, good point :) I'll probably do this as a GDScript initially, since i'm not confident enough to mess with the engine code right now.

**Comment by CiaraBurkett** [comment]
@QbieShay I agree with @Noshyaar, a dedicated node would enable people to use it for any number of things. And thanks for looking into this!

**Comment by QbieShay** [comment]
Hi, i started working on the spring node. For now it is really basic but i hope it solves some of the problems :) planning to make it a builtin node soon (TM).

https://github.com/QbieShay/GodotSpringArm

**Comment by QbieShay** [comment]
Hi, i made a first version of the [Spring Arm for Godot](https://github.com/QbieShay/GodotSpringArm). Right now, it position itself on the back of the target, using the target's up vector as up.

**Comment by QbieShay** [comment]
Hello everyone! After some time and some input, @Andrea and I implemented the spring arm node as a built in node. If you can, please check out this https://github.com/godotengine/godot/pull/18822 

Feedback is appreciated!
